worker/transfer.go

```diff
@@ -381,33 +429,29 @@ func parallelDownloadSlab(ctx context.Context, sp storeProvider, ss object.SlabS
 		}
 
 		if resp.err != nil {
-			errs = append(errs, &HostError{contracts[resp.req.hostIndex].HostKey, resp.err})
+			errs = append(errs, &HostError{hosts[resp.hostIndex].HostKey, resp.err})
 
 			// make sure non funded or gouging hosts are not used for consecutive downloads
 			if errors.Is(resp.err, errBalanceInsufficient) ||
 				errors.Is(resp.err, errGougingHost) {
-				timings[resp.req.hostIndex] = math.MaxInt64
+				timings[resp.hostIndex] = math.MaxInt64
 			}
 
 			// make sure slow hosts are not not used for consecutive downloads
 			if errors.Is(resp.err, errDownloadSectorTimeout) {
-				timings[resp.req.hostIndex] = int64(resp.dur) * 10
+				timings[resp.hostIndex] = int64(resp.dur) * 10
 			}
 
 			// try next host
-			if hostIndex < len(contracts) {
-				go worker(req{hostIndex, offset, length})
-				hostIndex++
-				inflight++
+			hostIndex := workerForShard(resp.req.shardIndex)
+			if hostIndex != -1 {
+				launchWorker(resp.req)
 			}
 		} else {
-			timings[resp.req.hostIndex] = int64(resp.dur)
-			for i := range ss.Shards {
-				if ss.Shards[i].Host == contracts[resp.req.hostIndex].HostKey && len(shards[i]) == 0 {
-					shards[i] = resp.shard
-					rem--
-					break
-				}
+			timings[resp.hostIndex] = int64(resp.dur)
+			if len(shards[resp.req.shardIndex]) == 0 {
+				shards[resp.req.shardIndex] = resp.shard
+				rem--
```

Comment:
the download code would be more consistent with the upload code if we remove the `rem > 0` from the loop condition and move it here right after decrementing the remaining shards and break if 0